build.rs

```diff
@@ -1,39 +1,83 @@
 // Copyright 2018 the Deno authors. All rights reserved. MIT license.
 
 // Run "cargo build -vv" if you want to see gn output.
-// TODO For the time being you must set an env var DENO_BUILD_PATH
-// which might be `pwd`/out/debug or `pwd`/out/release.
-// TODO Currently DENO_BUILD_PATH must be absolute.
-// TODO Combine DENO_BUILD_PATH and OUT_DIR.
+
+#![deny(warnings)]
 
 use std::env;
+use std::path::{self, Path, PathBuf};
 use std::process::Command;
 
 fn main() {
+  // Cargo sets PROFILE to either "debug" or "release", which conveniently
+  // matches the build modes we support.
   let mode = env::var("PROFILE").unwrap();
-  let deno_build_path = env::var("DENO_BUILD_PATH").unwrap();
+
+  // Normally we configure GN+Ninja to build into Cargo's OUT_DIR.
+  // However, when DENO_BUILD_PATH is set, perform the ninja build in that dir
+  // instead. This is used by CI to avoid building V8 etc twice.
+  let out_dir = env::var_os("OUT_DIR").unwrap();
+  let gn_out_dir = match env::var_os("DENO_BUILD_PATH") {
+    None => abs_path(out_dir),
+    Some(deno_build_path) => abs_path(deno_build_path),
+  };
+
+  // Give cargo some instructions. We do this first so the `rerun-if-*-changed`
+  // directives can take effect even if something the build itself fails.
+  println!("cargo:rustc-env=GN_OUT_DIR={}", gn_out_dir);
+  println!("cargo:rustc-link-search=native={}/obj", gn_out_dir);
+  println!("cargo:rustc-link-lib=static=deno_deps");
+
+  println!("cargo:rerun-if-changed={}", abs_path("src/msg.fbs"));
```

Comment:
This is getting dangerous. What is the behavior you’re trying to acheive / work around?
My goal is that this script be as dumb as possible and leave the build logic to our actual build scripts.

Reply:
It solved my intellisense not updating after modifying the spawn message :)

It's still very broken of course -- any other change to libdeno or v8 or whatever dependency is ignored by cargo; it won't run build.rs so ninja never even gets the chance to rebuild it.